enhancements/cluster-logging/forwarding-json-structured-logs.md

```diff
@@ -0,0 +1,476 @@
+* Elasticsearch indexing of JSON logs with multiple schema without index explosions.
+* Upgrade path from today's implementation.
+
+### Non-Goals
+
+* General-purpose JSON queries and transformations.
+* Recording or validating user-defined schema.
+  The forwarder only identifies schema by name, all schema knowledge belongs to the user.
+
+## Proposal
+
+#### Data model
+
+Two new top-level fields added to the record [data model][data_model]:
+
+* `structured` (type object): Original structured JSON log entry.
+* `schema` (type string): Schema name assigned by user.
+
+Relationship between `structured` and `message` fields:
+
+* EXACTLY ONE of `message` or `structured` MUST be present.
+* `schema` is OPTIONAL, defined and interpreted by the user.
+* The presence or absence of the `structured` field can be used to detect structured/unstructured records.
+* If a filter (see below) removes the `structured` field, an equivalent `message` field is added by.
+
+With an *Elasticsearch* store, `schema` is copied to the existing `viaq_index_name` field to set the index.
+`viaq_index_name` will continue to be used for compatibility and upgrade, but will eventually be removed.
+
+#### ClusterLogForwarder configuration
+
+The elements of the CLF API are:
+
+- *Input*: Selects logs, adds meta-data, creates _internal_ records.
+- *Pipeline*: Transforms, filters and routes internal records.
+- *Output*: Connects to consumers, encodes and sends _output_ records.
+
+New fields:
+
+`input.stuctured`: (string, default "json") Attempt to parse structured entries.
+The only legal value is "json", others may be added in future.
+
+An input record contains:
+
+* `structured` field if `structured=json` *and* parse succeeds.
+* `message` field if `input.structured` is absent *or* JSON parse fails.
+
+`input.schema`: (string, optional) User-defined schema name.
+Adds a `schema` field to the internal record with a user-defined value.
+Not used by the forwarder, included in output record for the consumer.
+
+`pipeline.filter`: (object, optional) Remove unwanted records.
+
+* `reject`: (array of string) drop records that have any of the listed fields.
+* `require`: (array of string) drop records that _do not_ have all the listed fields.
+
+`pipeline.modify`: (object, optional) Modify record fields.
+
+* `remove`: (array of string) remove named fields from the output record. \
+  **NOTE**: As a special case, removing the `structured` field replaces it with an equivalent unstructured `message` field.
+
+* `keep`: (array of string) keep only the named fields, remove all others.
+* `copy`: (object, optional, *NICE-TO-HAVE*) copies meta-data fields to the structured message.
+  Object keys are meta-data fields, values are destination fields relative to `structured`.
+  Copy is evaluated before `remove`, `keep` or `merge`
+* `merge`: (bool, optional,  *NICE-TO-HAVE*) If true, replace merge the fields of `structured` into the top-level record. \
```

Comment:
I think we should be very careful in allowing this transformation since we could overwrite fields in the entire payload from the message.  This opens us up to having cases where user is expecting a field we create but its being replaced by theirs which could be challenging and time consuming for us to debug.  Might we not just prefer to have various flavors of envelopes:
* RH provided meta (what we have now) which copies parsed messages into subkey
* User's entire message IS the payload
* Raw as-is

Reply:
Gotcha - so not "merge" just "replace". That makes sense to me, mixing the tow is potentially messy. I'll update. What's the difference between "users message IS the payload" and "raw as-is"?

Reply:
Hmm, that does eliminate the drop-in replacement for MERGE_JSON_LOG however. Do we need that?
@sichvoge

Reply:
+1 concern regarding `merge`

Reply:
Merge has been removed - not needed since you can achieve the same with `add` and `replace`, but now the onus is more clearly on the user to deal with name clashes. Also using `replace` will add field `@cluster_logging_modified@: true` to assist debugging if modified records get mixed up with cluster-logging data-model records.